adobe-platform/mesos-systemd
# docker initiated logrotation

Docker logs need to be rotated by the container instances themselves when using the default json logger because the container maintains an open file handle to the backing log file. From Coreos, this file can be found in /var/lib/docker/containers/_/_.log. To remove them from Cores though causes overlay and devicemapper corruption (unable to remove layers, images and or instances).
In summary,

All docker run commands have been provisioned with max-file & max-size. The values are pulled from etcd.
A default etcd value is set in init if the user doesn't set one via secrets.json
The external logrotate unit is removed.
The docker cleanup script is changed to use the docker client to remove dead containers and untagged layers i.e. those that show as '' when docker images is run.
Setting secrets.json type=etcd and key of /docker/config/logs-max-size and /docker/config/logs-max-file override the default of max-size=10m and max-file=10 i.e. no more than 10 rotated logs and of no more than 10MB in size.
Finally, the related ticket is adobe-community/issues#9054

Testing with ansible after convergence:
- Setup the [inventory](https://git.corp.adobe.com/fortescu/Mesos4Dexi/tree/master/inventory-ethos-f4tq)
- Checking the journals to make sure all `docker run` commands have targeted options

```
ansible coreos_control:coreos_workers -m raw -a 'bash -c "journalctl --no-pager | grep \"docker run\" | grep  -v max-size"' -i ~/Mesos4Dexi/inventory-ethos-f4tq/ -s -f 10
```
- Make sure systemctl ok

```
ansible coreos_control:coreos_workers:coreos_proxy -i ~/Mesos4Dexi/inventory-ethos-f4tq/  -m raw -a 'bash -c "systemctl list-units | grep -v active"'
```

which shouldn't list anything.  

> Note: remove `-v` to see all the appropriately decorated docker commands
- Test marathon access from each discrete node:

```
ansible coreos_control -m raw -a 'curl -L -v -u admin:password "http://{{ ec2_private_ip_address }}:8080"' -s -i ~/Mesos4Dexi/inventory-ethos-f4tq/ 
```

> Note: change the port/username/password for other well known processes 
- Check that all images are pulled on workers

```
ansible coreos_workers -m raw -a "docker images " -i ~/Mesos4Dexi/inventory-ethos-f4tq/ -s -f 10 -vvv 
```
- Check that all images are running on workers. 

```
ansible coreos_workers -m raw -a "docker ps " -i ~/Mesos4Dexi/inventory-ethos-f4tq/ -s -f 10 -vvv 
```
- Inspect last 20 lines of mesos master journal across your control nodes

```
ansible coreos_control -m raw -a 'bash -c "systemctl --no-pager list-units|grep master|grep service|awk \"{print \\\$1}\"| xargs -n 1 -IXX journalctl --no-pager -u XX | tail -20 "' -i ~/Mesos4Dexi/inventory-ethos-f4tq/ -s -f 10
```
- Inspect last 20 lines of mesos workers journal across your worker nodes

```
ansible coreos_workers -m raw -a 'bash -c "systemctl --no-pager list-units|grep slave|grep service|awk \"{print \\\$1}\"| xargs -n 1 -IXX journalctl --no-pager -u XX | tail -20 "' -i ~/Mesos4Dexi/inventory-ethos-f4tq/ -s -f 10
```

Anyhow, all useful for testing this PR.
